internal/configmodels/configmodels.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2019, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package configmodels defines the data models for entities. This file defines the
+// models for V2 configuration format. The defined entities are:
+// ConfigV2 (the top-level structure), Receivers, Exporters, Processors, Pipelines.
+package configmodels
+
+/*
+Receivers, Exporters and Processors typically have common configuration settings, however
+sometimes specific implementations (e.g. StackDriver) will have extra configuration settings.
+This requires the configuration data for these entities to be polymorphic.
+
+To satisfy these requirements we declare interfaces ReceiverCfg, ExporterCfg, ProcessorCfg,
+which define the behavior. We also provide helper structs ReceiverCommon, ExporterCommon,
+ProcessorCommon, which define the common settings and un-marshaling from config files.
+
+Specific Receivers/Exporters/Processors are expected to at the minimum implement the
+corresponding interface and if they have additional settings they must also extend
+the corresponding common settings struct (the easiest approach is to embed the common struct).
+*/
+
+// ConfigV2 defines the configuration V2 for the various elements of collector or agent.
+type ConfigV2 struct {
+	Receivers  []ReceiverCfg  `mapstructure:"-"`
```

Comment:
Not sure why for Receivers we use []ReceiverCfg but for Pipeline we use Pipelines.

Should we have consistency here and add a Receivers struct similar with Pipelines? Also that will make the mapstructure clearer.

Reply:
>Not sure why for Receivers we use []ReceiverCfg but for Pipeline we use Pipelines.

You are right, for naming consistency this is better named PipelinesCfg. I will change.

>Should we have consistency here and add a Receivers struct similar with Pipelines? Also that will make the mapstructure clearer.

We need `Pipelines` to be a separate struct because it is contains sub-keys (traces and metrics) and is used as an argument to viper.Unmarshal().

For Receivers/Exporters/Processors we do not have such need, they are unmarshaled manually and are an array of items, not a struct. I do not have a strong opinion, perhaps we can make these structs too, I do not think it will make significant difference.